File: bmsspy/quicksplit.py
def median(arr:list[int|float]) -> int|float:
    """
    Function:

    - Calculates the median of a list of numbers by sorting the list and finding the middle value or the average of the two middle values.

    Requires:

    - arr: A list of integers or floats.

    Returns:

    - The median value as an integer or float.
    """
    len_arr = len(arr)
    idx = len_arr // 2
    sorted_arr = sorted(arr)
    if len_arr%2 == 1:
        return sorted_arr[idx]
    else:
        return (sorted_arr[idx - 1] + sorted_arr[idx]) / 2

def median_of_medians(arr:list[int|float], split_size:int=5) -> int|float:
    """
    Function:

    - Computes the median of medians of a list of numbers. 
    - This is done by dividing the list into sublists of a fixed size (5 in this case)
        - For each sublist, find the median of each sublist
        - Then iteratively find the median of those medians
        - Stop when the list is smaller than or equal to the split size and return the median of that list

    Required Arguments:

    - arr: A list of integers or floats.

    Optional Arguments:

    - split_size: The size of the sublists to create
        - Default is 5
        - Must be an odd number to ensure a single median value

    """
    split_median_idx = split_size // 2

    while True:
        len_arr = len(arr)
        if len_arr <= split_size:
            return median(arr)
        extra = []
        # Allow for arrays not divisible by split_size
        if len_arr % split_size != 0:
            extra = [median(arr[len_arr - (len_arr % split_size):])]
            arr = arr[:len_arr - (len_arr % split_size)]
        medians = [sorted(arr[i:i+split_size])[split_median_idx] for i in range(0, len(arr), split_size)]
        arr = medians + extra

def quicksplit(arr, lower_bucket_size:int=None):
    """
    Function:

    - Splits an array into two buckets using a variant of the Quickselect algorithm.

    Required Arguments:

    - arr: A list of integers or floats to be split.

    Optional Arguments:

    - lower_bucket_size: The desired size of the lower bucket.
        - If not provided, the function will split the array into two equal halves (or as close as possible).
    
    Returns:

    - A dictionary with two keys:
        - 'lower': A list containing the lower bucket of elements.
        - 'higher': A list containing the higher bucket of elements.
    """
    # If no lower bucket size is given, split in half or as close as possible
    if lower_bucket_size is None:
        lower_bucket_size = len(arr)//2
    higher = []
    lower = []
    while True:
        pivot = median_of_medians(arr)
        # Loop over the array once to partition into three lists
        # This is faster than using list 3 list comprehensions
        below = []
        pivots = []
        above = []
        for x in arr:
            if x < pivot:
                below.append(x)
            elif x > pivot:
                above.append(x)
            else:
                pivots.append(x)

        count_below = len(below) + len(lower)
        if lower_bucket_size < count_below:
            higher = pivots + above + higher
            arr = below
        elif lower_bucket_size > count_below + len(pivots):
            lower = lower + below + pivots
            arr = above
        else:
            pivot_split_idx = lower_bucket_size - count_below
            lower = lower + below + pivots[:pivot_split_idx]
            higher = pivots[pivot_split_idx:] + above + higher
            return {
                'lower': lower,
                'higher': higher
            }

def sortsplit(arr, lower_bucket_size:int=None):
    """
    Function:

    - Splits an array into two buckets by sorting the array and dividing it at a specified index.

    Required Arguments:

    - arr: A list of integers or floats to be split.

    Optional Arguments:

    - lower_bucket_size: The desired size of the lower bucket.
        - If not provided, the function will split the array into two equal halves (or as close as possible).
    
    Returns:

    - A dictionary with two keys:
        - 'lower': A list containing the lower bucket of elements.
        - 'higher': A list containing the higher bucket of elements.

    """
    if lower_bucket_size is None:
        lower_bucket_size = len(arr)//2
    sorted_arr = sorted(arr)
    return {
        'lower': sorted_arr[:lower_bucket_size],
        'higher': sorted_arr[lower_bucket_size:]
    }

File: cpp/src/quicksplit.cpp
#include "quicksplit.hpp"  // include header for declarations

#include <vector>
#include <algorithm>
#include <stdexcept>
#include <cmath>
#include <map>
#include <limits>

float median(std::vector<float> arr, bool split) {
    size_t len = arr.size();
    if (len == 0) throw std::invalid_argument("Cannot find median of empty array");

    std::sort(arr.begin(), arr.end());
    size_t idx = len / 2;

    if (len % 2 == 1) {
        return arr[idx];
    } else {
        if (split) {
            return (arr[idx - 1] + arr[idx]) / 2.0f;
        } else {
            return arr[idx - 1];
        }
    }
}

float median_of_medians(std::vector<float> arr, int split_size, bool split) {
    if (split_size % 2 == 0) {
        throw std::invalid_argument("split_size must be an odd number");
    }

    int split_median_idx = split_size / 2;

    while (true) {
        size_t len = arr.size();
        if (len <= static_cast<size_t>(split_size)) {
            return median(arr, split);
        }

        std::vector<float> medians;
        std::vector<float> extra;

        if (len % split_size != 0) {
            std::vector<float> tail(arr.end() - (len % split_size), arr.end());
            extra.push_back(median(tail));
            arr.resize(len - (len % split_size));
        }

        for (size_t i = 0; i < arr.size(); i += split_size) {
            std::vector<float> chunk(arr.begin() + i, arr.begin() + i + split_size);
            std::sort(chunk.begin(), chunk.end());
            medians.push_back(chunk[split_median_idx]);
        }

        arr = medians;
        arr.insert(arr.end(), extra.begin(), extra.end());
    }
}

QuickSplitResult quicksplit(std::vector<float> arr, int lower_bucket_size) {
    if (lower_bucket_size == -1) {
        lower_bucket_size = static_cast<int>(std::ceil(arr.size() / 2.0));
    }

    if (lower_bucket_size <= 0 || static_cast<size_t>(lower_bucket_size) > arr.size()) {
        throw std::invalid_argument("lower_bucket_size must be > 0 and <= arr size");
    }

    std::vector<float> lower, higher;

    while (true) {
        float pivot = median_of_medians(arr, 5, false);

        std::vector<float> below, above, pivots;
        for (float x : arr) {
            if (x < pivot)
                below.push_back(x);
            else if (x > pivot)
                above.push_back(x);
            else
                pivots.push_back(x);
        }

        int count_below = static_cast<int>(below.size() + lower.size());

        if (lower_bucket_size < count_below) {
            higher.insert(higher.end(), pivots.begin(), pivots.end());
            higher.insert(higher.end(), above.begin(), above.end());
            arr = below;
        } else if (lower_bucket_size > count_below + static_cast<int>(pivots.size())) {
            lower.insert(lower.end(), below.begin(), below.end());
            lower.insert(lower.end(), pivots.begin(), pivots.end());
            arr = above;
        } else {
            int pivot_split_idx = lower_bucket_size - count_below;
            lower.insert(lower.end(), below.begin(), below.end());
            lower.insert(lower.end(), pivots.begin(), pivots.begin() + pivot_split_idx);
            higher.insert(higher.end(), pivots.begin() + pivot_split_idx, pivots.end());
            higher.insert(higher.end(), above.begin(), above.end());

            float final_pivot;
            if (pivot_split_idx == 0 && !below.empty()) {
                final_pivot = *std::max_element(below.begin(), below.end());
            } else {
                final_pivot = pivot;
            }

            return {lower, higher, final_pivot};
        }
    }
}

QuickSplitDictResult quicksplit_dict(const std::map<int, float>& data, int lower_bucket_size) {
    if (lower_bucket_size == -1) {
        lower_bucket_size = static_cast<int>(std::ceil(data.size() / 2.0));
    }

    if (lower_bucket_size <= 0 || static_cast<size_t>(lower_bucket_size) > data.size()) {
        throw std::invalid_argument("lower_bucket_size must be > 0 and <= data size");
    }

    using Pair = std::pair<int, float>;
    std::vector<Pair> lower, higher;
    std::vector<Pair> arr(data.begin(), data.end());

    while (true) {
        std::vector<float> values;
        for (const auto& [key, val] : arr) {
            values.push_back(val);
        }

        float pivot = median_of_medians(values, 5, false);

        std::vector<Pair> below, above, pivots;
        for (const auto& item : arr) {
            if (item.second < pivot)
                below.push_back(item);
            else if (item.second > pivot)
                above.push_back(item);
            else
                pivots.push_back(item);
        }

        int count_below = static_cast<int>(below.size() + lower.size());

        if (lower_bucket_size < count_below) {
            higher.insert(higher.end(), pivots.begin(), pivots.end());
            higher.insert(higher.end(), above.begin(), above.end());
            arr = below;
        } else if (lower_bucket_size > count_below + static_cast<int>(pivots.size())) {
            lower.insert(lower.end(), below.begin(), below.end());
            lower.insert(lower.end(), pivots.begin(), pivots.end());
            arr = above;
        } else {
            int pivot_split_idx = lower_bucket_size - count_below;
            lower.insert(lower.end(), below.begin(), below.end());
            lower.insert(lower.end(), pivots.begin(), pivots.begin() + pivot_split_idx);
            higher.insert(higher.end(), pivots.begin() + pivot_split_idx, pivots.end());
            higher.insert(higher.end(), above.begin(), above.end());

            float final_pivot;
            if (pivot_split_idx == 0 && !below.empty()) {
                auto max_it = std::max_element(below.begin(), below.end(),
                    [](const Pair& a, const Pair& b) { return a.second < b.second; });
                final_pivot = max_it->second;
            } else {
                final_pivot = pivot;
            }

            std::map<int, float> lower_map, higher_map;
            for (const auto& item : lower) lower_map[item.first] = item.second;
            for (const auto& item : higher) higher_map[item.first] = item.second;

            return {lower_map, higher_map, final_pivot};
        }
    }
}
